### PR TITLE
feat: added background image

### PR DESCRIPTION
**Pull request by eliseeborn**

### CODE REVIEW

File: src/ext.js [hunk 1]
```diff
@@ -132,7 +149,89 @@ export default function ext(/* env */) {
                     color: null,
                   },
                 },
-              ],
+                useBackgroundImage: {
+                  ref: 'style.background.isUsed',
+                  type: 'boolean',
+                  translation: 'properties.backgroundImage.use',
+                  component: 'switch',
+                  defaultValue: false,
```
Comment (niekvanstaveren) [inline src/ext.js]: @Caele Do you think we will ever have the option to set a background image from the Theme? In that case we cannot set a defaultValue here, but it feels like quite a stretch

**Comment by niekvanstaveren** [comment]
@eliseeborn When I switched between fit to width and fit to height I see some errors popping up. Could you have a look at that?

**Comment by eliseeborn** [comment]
@niekvanstaveren So, it's due to that align-matrix assumes that background.position is in $scope.data, and here it's in like layout.style.background.position - so when align-matrix tries to change position because the old position is no longer allowed due to changing the size, it throws an error and doesn't change the position.
I'm not 100% sure about the structure of the scope, so I don't really know how to fix this.

**Comment by haxxmaxx** [comment]
Since this is open source we want to keep out internal stuff like jira numbers (Miralem told us yesterday, thats why there are older commits with jira numbers). When working in the client it's great to add them tho 👍

**Comment by eliseeborn** [comment]
ok, I'll remove it when I merge then :)

**Comment by haxxmaxx** [comment]
Great, change the name of the PR as well

File: src/ext.js (hunk 1 above)
Comment (Caele) [reply src/ext.js]: no that would be strange, the images right now has to come from within sense, you cannot link to an external one (which a theme one would almost be)

File: src/ext.js [hunk 2]
```diff
@@ -5,6 +5,23 @@ export const checkShow = (data, field) => {
   return act && act.requiredInput && act.requiredInput.indexOf(field) !== -1;
 };
 
+export const getValue = (data, reference, defaultValue) => {
+  const steps = reference.split('.');
+  let dataContainer = data;
```
Comment (haxxmaxx) [inline src/ext.js]: Just call the argument `dataContainer`?

Comment (eliseeborn) [reply src/ext.js]: it's a util function from client originally, but it doesn't seem to make any difference

Comment (niekvanstaveren) [reply src/ext.js]: Changing the variable names might make it easier to understand. Also I would suggest breaking this out into another file so we keep ext.js cleaner. (Yes that probably also goes for checkShow and another one that I am adding for navigation :))

File: src/ext.js [hunk 3]
```diff
@@ -126,16 +142,98 @@ export default function ext(/* env */) {
...
+                  show(data) {
+                    return (
+                      getValue(data, 'style.background.isUsed')
+                      && getValue(data, 'style.background.url.qStaticContentUrlDef.qUrl')
+                      && getValue(data, 'style.background.size') !== 'fill'
+                    );
+                  },
+                  currentSize(data) {
+                    return getValue(data, 'style.background.size');
+                  },
```
Comment (niekvanstaveren) [inline src/ext.js]: What does this one do?

File: src/utils/__tests__/style-formatter.spec.js [hunk 4]
```diff
@@ -42,5 +43,55 @@ describe('style-formatter', () => {
...
+          isUsed: true,
+          url: {
+            qStaticContentUrl: {
+              qUrl: someUrl
+            }
+          }
+        }
+      };
+      const formattedStyle = styleFormatter.getStyles(style, Theme);
+      expect(formattedStyle.includes(`background-image: url('http://localhost:4848${someUrl}')`)).to.be.true;
```
Comment (niekvanstaveren) [inline src/utils/__tests__/style-formatter.spec.js]: Interesting that this works for the pipeline as well

File: src/utils/style-formatter.js [hunk 5]
```diff
@@ -13,10 +33,22 @@ export default {
...
     palette = themeResolver.getPalette(Theme);
     styles += formatColorProperty('color', style.fontColor, '#ffffff');
     styles += formatProperty('font-size', style.fontSize ? `${style.fontSize}px` : '12px');
     styles += formatColorProperty('background-color', style.backgroundColor, '#3F8AB3');
+    if (style && style.background && style.background.isUsed) {
+      let bgUrl = style.background.url.qStaticContentUrl.qUrl;
+      bgUrl.replace('http://localhost:4848/', '').replace(/^\.\.\//i, '/');
+      bgUrl = bgUrl.replace(/"/g, '\\"');
+      bgUrl = bgUrl.replace(/'/g, "\\'");
+      const bgImg = `http://localhost:4848/${(bgUrl[0] === '/' ? bgUrl.substr(1) : bgUrl)}`;
```
Comment (niekvanstaveren) [inline src/utils/style-formatter.js]: This won't work for environments not running on localhost

File: src/ext.js (hunk 3 above)
Comment (eliseeborn) [reply src/ext.js]: i think this is what align-matrix uses to watch the changes in size?

File: src/utils/style-formatter.js [hunk 6]
```diff
@@ -13,10 +33,21 @@ export default {
...
+
     palette = themeResolver.getPalette(Theme);
     styles += formatColorProperty('color', style.fontColor, '#ffffff');
     styles += formatProperty('font-size', style.fontSize ? `${style.fontSize}px` : '12px');
     styles += formatColorProperty('background-color', style.backgroundColor, '#3F8AB3');
+    if (style && style.background && style.background.isUsed) {
+      let bgUrl = style.background.url.qStaticContentUrl.qUrl;
+      bgUrl.replace(/^\.\.\//i, '/');
+      bgUrl = bgUrl.replace(/"/g, '\\"');
+      bgUrl = bgUrl.replace(/'/g, "\\'");
```
Comment (niekvanstaveren) [inline src/utils/style-formatter.js]: @Caele I think these checks were copied from the text and image implementation. What we don't have here is the addition of the base path url. Would this be a requirement? Also more in general how do we make sure we are good on a security perspective?

**Pull request merged**